### PR TITLE
Compute when to sync epoch start

### PR DESCRIPTION
**Pull request by sasurobert**

### CODE REVIEW

File: epochStart/bootstrap/process.go [hunk 1]
```diff
@@ -181,27 +186,21 @@ func NewEpochStartBootstrap(args ArgsEpochStartBootstrap) (*epochStartBootstrap,
...
 	if isCurrentTimeBeforeGenesis {
 		return true
 	}
 
-	configuredDurationOfEpoch := startTime.Add(e.computedDurationOfEpoch())
-	isEpochZero := time.Since(configuredDurationOfEpoch) < 0
-
-	return isEpochZero
+	currentRound := e.rounder.Index()
+	return float64(currentRound) < float64(e.generalConfig.EpochStartConfig.RoundsPerEpoch)*gracePeriodInPercentage
```
Comment (iulianpascalau) [inline epochStart/bootstrap/process.go:203]: this won't work with the early start of epoch. Can remain like this for time being.

File: epochStart/bootstrap/process.go [hunk 2]
```diff
@@ -39,9 +39,12 @@ import (
 
 var log = logger.GetOrCreate("epochStart/bootstrap")
 
-const timeToWait = 10 * time.Second
+const timeToWait = time.Minute
+const trieSyncWaitTime = 5 * time.Minute
 const timeBetweenRequests = 100 * time.Millisecond
 const maxToRequest = 100
+const gracePeriodInPercentage = float64(0.25)
```
Comment (SebastianMarian) [inline epochStart/bootstrap/process.go:46]: This is equivalent to ShuffledOutRestartThreshold from config.toml [EpochStartConfig] section ?

File: epochStart/bootstrap/process.go [hunk 3]
```diff
@@ -285,6 +268,21 @@ func (e *epochStartBootstrap) Bootstrap() (Parameters, error) {
 	return e.requestAndProcessing()
 }
 
+func (e *epochStartBootstrap) computeIfCurrentEpochIsSaved() bool {
+	e.initializeFromLocalStorage()
```
Comment (iulianpascalau) [inline epochStart/bootstrap/process.go:272]: We have a bug in the method initializeFromLocalStorage that can appear if we have an epoch DB saved but on the wrong shard. So, the method `getParentDirAndLastEpoch` should take into account also the shard ID

File: epochStart/bootstrap/process.go (hunk 1 above)
Comment (SebastianMarian) [inline epochStart/bootstrap/process.go:203]: We don't need to use here MinRoundsBetweenEpochs from config.toml [EpochStartConfig] section instead 
float64(e.generalConfig.EpochStartConfig.RoundsPerEpoch)*gracePeriodInPercentage ?

File: epochStart/bootstrap/process.go [hunk 4]
```diff
@@ -285,6 +268,21 @@ func (e *epochStartBootstrap) Bootstrap() (Parameters, error) {
...
+		return false
+	}
+
+	computedRound := e.rounder.Index()
+	if computedRound-e.baseData.lastRound < roundGracePeriod {
+		return true
+	}
+
+	roundsSinceEpochStart := computedRound - int64(e.baseData.epochStartRound)
+	return float64(roundsSinceEpochStart) < float64(e.generalConfig.EpochStartConfig.RoundsPerEpoch)*gracePeriodInPercentage
```
Comment (SebastianMarian) [inline epochStart/bootstrap/process.go:283]: We don't need to use here MinRoundsBetweenEpochs from config.toml [EpochStartConfig] section instead
float64(e.generalConfig.EpochStartConfig.RoundsPerEpoch)*gracePeriodInPercentage ?

File: epochStart/bootstrap/process.go [hunk 5]
```diff
@@ -285,6 +268,21 @@ func (e *epochStartBootstrap) Bootstrap() (Parameters, error) {
...
 }
 
+func (e *epochStartBootstrap) computeIfCurrentEpochIsSaved() bool {
+	e.initializeFromLocalStorage()
+	if !e.baseData.storageExists {
+		return false
+	}
+
+	computedRound := e.rounder.Index()
+	if computedRound-e.baseData.lastRound < roundGracePeriod {
```
Comment (SebastianMarian) [inline epochStart/bootstrap/process.go:278]: Is this condition safe, we really need it?

File: storage/factory/openStorage.go [hunk 6]
```diff
@@ -271,37 +284,85 @@ func getLastEpochAndRoundFromStorage(
...
-		bootstrapData, errGet := getBootstrapDataForPersisterPath(persisterFactory, persisterPath, marshalizer)
+		bootstrapData, storer, errGet := getBootstrapDataForPersisterPath(persisterFactory, persisterPath, marshalizer)
 		if errGet != nil {
 			continue
 		}
 
 		if bootstrapData.LastRound > highestRoundInStoredShards {
+			shardID := uint32(0)
+			var err error
+			shardID, err = convertShardIDToUint32(shardIdStr)
```
Comment (SebastianMarian) [inline storage/factory/openStorage.go:295]: Why not shardID, errConvert := and you can remove both lines above

File: storage/factory/openStorage.go [hunk 7]
```diff
@@ -271,37 +284,85 @@ func getLastEpochAndRoundFromStorage(
...
+			if err != nil {
+				continue
```
Comment (SebastianMarian) [inline storage/factory/openStorage.go:297]: Maybe we can add a log here

File: storage/factory/openStorage.go [hunk 8]
```diff
@@ -271,37 +284,85 @@ func getLastEpochAndRoundFromStorage(
...
+			}
+			epochStartRound, err = loadEpochStartRound(shardID, bootstrapData.EpochStartTriggerConfigKey, storer)
+			if err != nil {
+				continue
```
Comment (SebastianMarian) [inline storage/factory/openStorage.go:301]: Maybe we can add a log here

File: epochStart/bootstrap/process.go (hunk 3 above)
Comment (sasurobert) [reply epochStart/bootstrap/process.go:272]: @bogdan-rosianu is working on refactoring openStorage.go file .

File: cmd/node/main.go [hunk 9]
```diff
@@ -664,6 +665,15 @@ func startNode(ctx *cli.Context, log logger.Logger, version string) error {
 		return err
 	}
 
+	rounder, err := round.NewRound(
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply cmd/node/main.go:668]: in epoch start in does not do much. It is a simple component, I do not see it problematic of being duplicated.

File: epochStart/bootstrap/process.go (hunk 1 above)
Comment (sasurobert) [reply epochStart/bootstrap/process.go:203]: yep. but that is hard to know in this case.

File: cmd/node/main.go (hunk 9 above)
Comment (SebastianMarian) [inline cmd/node/main.go:668]: Please pass this rounder to NewProcessComponentsFactoryArgs right below in line 906, and remove its second creation from func ProcessComponentsFactory(args *processComponentsFactoryArgs) (*Process, error), to have only one instance of rounder in the whole project

File: epochStart/bootstrap/process.go (hunk 1 above)
Comment (sasurobert) [reply epochStart/bootstrap/process.go:203]: MinRoundsBetweenEpoch is used only for forceEndOfEpoch.

File: epochStart/bootstrap/process.go (hunk 4 above)
Comment (sasurobert) [reply epochStart/bootstrap/process.go:283]: MinRoundsBetweenEpoch is used only for forceEndOfEpoch.

File: epochStart/bootstrap/process.go (hunk 5 above)
Comment (sasurobert) [reply epochStart/bootstrap/process.go:278]: the condition is safe. I just added if for any case the round is relatively close to one in the DB there is no reason to start syncing.

File: epochStart/bootstrap/process.go (hunk 2 above)
Comment (sasurobert) [reply epochStart/bootstrap/process.go:46]: right now yes. but I do not know what gracePeriod should we choose.

File: storage/factory/openStorage.go (hunk 6 above)
Comment (sasurobert) [reply storage/factory/openStorage.go:295]: done

File: storage/factory/openStorage.go (hunk 7 above)
Comment (sasurobert) [reply storage/factory/openStorage.go:297]: done

File: storage/factory/openStorage.go (hunk 8 above)
Comment (sasurobert) [reply storage/factory/openStorage.go:301]: done

File: cmd/node/main.go (hunk 9 above)
Comment (sasurobert) [reply cmd/node/main.go:668]: done

File: cmd/node/factory/structs.go [hunk 10]
```diff
@@ -548,6 +547,7 @@ type processComponentsFactoryArgs struct {
 	nodesConfig               *sharding.NodesSetup
 	gasSchedule               map[string]map[string]uint64
 	syncer                    ntp.SyncTimer
```
Comment (SebastianMarian) [inline cmd/node/factory/structs.go:549]: Delete syncer var as it is unused now

File: epochStart/bootstrap/process.go (hunk 1 above)
Comment (iulianpascalau) [reply epochStart/bootstrap/process.go:203]: agree, probably we will have a dedicated component for this

File: epochStart/bootstrap/process.go (hunk 3 above)
Comment (iulianpascalau) [reply epochStart/bootstrap/process.go:273]: great!

File: epochStart/bootstrap/process.go (hunk 2 above)
Comment (SebastianMarian) [reply epochStart/bootstrap/process.go:46]: I think we should add a TODO here to not forget later to replace these constants, gracePeriodInPercentage and roundGracePeriod, with values read from config.toml

File: epochStart/bootstrap/process.go (hunk 1 above)
Comment (SebastianMarian) [reply epochStart/bootstrap/process.go:203]: Yes I know, and I just thinking to this case, just to work even in this situation. But anyway, you can leave it like this right now, if this is not a good suggestion

File: epochStart/bootstrap/process.go (hunk 4 above)
Comment (SebastianMarian) [reply epochStart/bootstrap/process.go:287]: Yes I know, and I just thinking to this case, just to work even in this situation. But anyway, you can leave it like this right now, if this is not a good suggestion

File: epochStart/bootstrap/process.go (hunk 5 above)
Comment (SebastianMarian) [reply epochStart/bootstrap/process.go:280]: Agree, but probably would be catch anyway by the next computation

File: epochStart/bootstrap/process.go (hunk 2 above)
Comment (sasurobert) [reply epochStart/bootstrap/process.go:46]: I think this should not be configurable. we have to find the best values.

File: epochStart/bootstrap/process.go (hunk 1 above)
Comment (sasurobert) [reply epochStart/bootstrap/process.go:203]: it is impossible to know at start in epoch if there was a force start in epoch or not.

File: epochStart/bootstrap/process.go (hunk 4 above)
Comment (sasurobert) [reply epochStart/bootstrap/process.go:287]: t is impossible to know at start in epoch if there was a force start in epoch or not.

File: cmd/node/factory/structs.go (hunk 10 above)
Comment (sasurobert) [reply cmd/node/factory/structs.go:549]: done.

File: storage/factory/openStorage.go [hunk 11]
```diff
@@ -365,17 +382,22 @@ func getBootstrapDataForPersisterPath(
 		return nil, nil, err
 	}
 
-	defer func() {
-		errClose := persister.Close()
```
Comment (iulianpascalau) [inline storage/factory/openStorage.go:308]: This was really wrong

**Pull request merged**